### PR TITLE
CC-7633: ConfigDef needs to check current thread's classloader when parsing a type CLASS config

### PR DESCRIPTION
**Pull request by kkonstantine**
After KAFKA-8340, KAFKA-8819 were addressed in apache/kafka this ConfigDef class needs also to consider the current thread's classloader when parsing the type CLASS config type. 

This brings parsing for these types closer to the equivalent class in apache/kafka: org.apache.kafka.common.config.ConfigDef#parseType

### CODE REVIEW

**Comment by kkonstantine** [comment]
Working on a unit test. Will update PR as soon as I figure out viability.

**Comment by kkonstantine** [comment]
Added the necessary test. This is ready for review on my side.

File: config/src/test/java/io/confluent/common/config/ConfigDefTest.java [hunk 1]
```diff
@@ -275,4 +285,79 @@ private void testValidators(Type type, ConfigDef.Validator validator, Object def
...
+    Set<String> allowedKlasses = Collections.singleton(configNameDefault);
+    List<URL> pluginUrls = Collections.singletonList(Paths.get("target/classes/").toUri().toURL());
+    URLClassLoader pluginLoader = new PluginClassLoader(pluginUrls.toArray(new URL[0]), systemLoader, allowedKlasses);
+
+    Thread.currentThread().setContextClassLoader(pluginLoader);
+
+    configDef = new ConfigDef();
+    configDef.define(configName, Type.CLASS, configNameDefault, ConfigDef.Importance.MEDIUM, "plugin class");
+
+    config = new AbstractConfig(configDef, new HashMap<>());
```
Comment (rhauch) [inline config/src/test/java/io/confluent/common/config/ConfigDefTest.java]: Would it make sense to go a bit further and try getting the class from `config.getClass(configName)`, both here and above?

File: config/src/test/java/io/confluent/common/config/ConfigDefTest.java [hunk 2]
```diff
@@ -275,4 +285,79 @@ private void testValidators(Type type, ConfigDef.Validator validator, Object def
...
+
+    Class<?> systemKlass = config.getClass(configName);
+    assertEquals(systemLoader, systemKlass.getClassLoader());
+
+    // test with custom, child-first, classloader
+    Set<String> allowedKlasses = Collections.singleton(configNameDefault);
+    List<URL> pluginUrls = Collections.singletonList(Paths.get("target/classes/").toUri().toURL());
+    URLClassLoader pluginLoader = new PluginClassLoader(pluginUrls.toArray(new URL[0]), systemLoader, allowedKlasses);
+
+    Thread.currentThread().setContextClassLoader(pluginLoader);
```
Comment (gharris1727) [inline config/src/test/java/io/confluent/common/config/ConfigDefTest.java]: I think you should swap back out of this loader at the end of the test (or in a finally block).

File: config/src/test/java/io/confluent/common/config/ConfigDefTest.java [hunk 3]
```diff
@@ -275,4 +285,79 @@ private void testValidators(Type type, ConfigDef.Validator validator, Object def
...
+    Class<?> pluginKlass = config.getClass(configName);
+    assertEquals(pluginLoader, pluginKlass.getClassLoader());
+    assertNotEquals(pluginKlass, systemKlass);
+    assertEquals(pluginKlass.getName(), systemKlass.getName());
+  }
+
+  // Reduced dummy implementation of a child-first classloader based on
+  // org.apache.kafka.connect.runtime.isolation.PluginClassLoader
+  // Should be used only for testing
+  public class PluginClassLoader extends URLClassLoader {
```
Comment (gharris1727) [inline config/src/test/java/io/confluent/common/config/ConfigDefTest.java]: I would make this private and/or change the name to make sure that this is never confused for the real plugin class loader.
```suggestion
  private class SimplePluginClassLoader extends URLClassLoader {
```

File: config/src/test/java/io/confluent/common/config/ConfigDefTest.java (hunk 1 above)
Comment (kkonstantine) [reply config/src/test/java/io/confluent/common/config/ConfigDefTest.java]: Not sure I follow the suggestion. 

`config.getClass(configName)` is called indeed twice. 
Once in line 303 and once in 318 (current line numbers). 

If you are referring to `configNameDefault = Password.class.getName();` probably better to not call this to leave it to the ConfigDef parsing to load this class when it constructs the defintion of the config. (Doesn't make a difference if we add this call before the definition).

**Comment by rhauch** [comment]
LGTM. Thanks!

**Comment by kkonstantine** [comment]
Thanks for the review @rhauch !
I replied to your comment and added a minor commit to remove the unused imports from the test (that don't break checkstyle)

File: config/src/test/java/io/confluent/common/config/ConfigDefTest.java (hunk 1 above)
Comment (rhauch) [reply config/src/test/java/io/confluent/common/config/ConfigDefTest.java]: I missed that. Apologies.

File: config/src/test/java/io/confluent/common/config/ConfigDefTest.java [hunk 4]
```diff
@@ -275,4 +285,79 @@ private void testValidators(Type type, ConfigDef.Validator validator, Object def
...
+
+    // test with custom, child-first, classloader
+    Set<String> allowedKlasses = Collections.singleton(configNameDefault);
+    List<URL> pluginUrls = Collections.singletonList(Paths.get("target/classes/").toUri().toURL());
+    URLClassLoader pluginLoader = new PluginClassLoader(pluginUrls.toArray(new URL[0]), systemLoader, allowedKlasses);
+
+    Thread.currentThread().setContextClassLoader(pluginLoader);
+
+    configDef = new ConfigDef();
+    configDef.define(configName, Type.CLASS, configNameDefault, ConfigDef.Importance.MEDIUM, "plugin class");
```
Comment (gharris1727) [inline config/src/test/java/io/confluent/common/config/ConfigDefTest.java]: I think you can re-use the same ConfigDef object, and just have the AbstractConfig constructed with the alternate classloader.

**Comment by gharris1727** [comment]
Also is this fix only applicable to 5.3.x, or can it be backported farther? AK#6959 and AK#7315 were both backported to 1.0

File: config/src/test/java/io/confluent/common/config/ConfigDefTest.java (hunk 4 above)
Comment (kkonstantine) [reply config/src/test/java/io/confluent/common/config/ConfigDefTest.java]: I construct the `AbstractConfig` in order to get the value back. My test as intended tests `parseType` inside `define`. I don't see a good reason to add a redirection here. Also, I think I would have to add the type in the map in order to call `parseType` during `AbstractConfig`'s creation. Again, given that I don't want to change visibility of methods in this PR, I'd like to keep this test as close as possible to a unit test of the parsing.

File: config/src/test/java/io/confluent/common/config/ConfigDefTest.java (hunk 2 above)
Comment (kkonstantine) [reply config/src/test/java/io/confluent/common/config/ConfigDefTest.java]: Good idea.

File: config/src/test/java/io/confluent/common/config/ConfigDefTest.java (hunk 3 above)
Comment (kkonstantine) [reply config/src/test/java/io/confluent/common/config/ConfigDefTest.java]: Added `protected`. Should have the same effect.

File: config/src/test/java/io/confluent/common/config/ConfigDefTest.java (hunk 4 above)
Comment (gharris1727) [reply config/src/test/java/io/confluent/common/config/ConfigDefTest.java]: Oh I see, the critical parseType is actually in `define` in the confluent ConfigDef, which is different than AK's ConfigDef. This makes my earlier suggestion invalid.

**Comment by kkonstantine** [comment]
Thanks @gharris1727 
Good observation regarding backporting. The actual fixes were backported to AK 0.11.0.4
I rebased the changes after compiling against Java 7 on `3.3.x`. There a minor conflict on a single import statement in the test class which I'll resolve when merging.

**Comment by kkonstantine** [comment]
Turns out that this minor conflict confuses jenkins. I'll squash and merge on the initial target branch  here (`5.3.x`) and will cherry-pick to previous branches later on.

**Comment by kkonstantine** [comment]
Yeah, ticket creation is implied. Will be addressed in CC-7646
Thanks

**Pull request merged**